agent/hcp/client/metrics_client.go

```diff
@@ -0,0 +1,158 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"golang.org/x/oauth2"
+	"google.golang.org/protobuf/proto"
+
+	colmetricpb "go.opentelemetry.io/proto/otlp/collector/metrics/v1"
+	metricpb "go.opentelemetry.io/proto/otlp/metrics/v1"
+
+	"github.com/hashicorp/consul/version"
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-retryablehttp"
+	hcpcfg "github.com/hashicorp/hcp-sdk-go/config"
+)
+
+const (
+	// HTTP Client config
+	defaultStreamTimeout = 15 * time.Second
+
+	// Retry config
+	defaultRetryWaitMin = 1 * time.Second
+	defaultRetryWaitMax = 15 * time.Second
+	defaultRetryMax     = 4
+)
+
+// MetricsClient exports Consul metrics in OTLP format to the HCP Telemetry Gateway.
+type MetricsClient interface {
+	ExportMetrics(ctx context.Context, protoMetrics *metricpb.ResourceMetrics, endpoint string) error
+}
+
+// cloudConfig represents cloud config for TLS abstracted in an interface for easy testing.
+type cloudConfig interface {
+	HCPConfig(opts ...hcpcfg.HCPConfigOption) (hcpcfg.HCPConfig, error)
+}
+
+// otlpClient is an implementation of MetricsClient with a retryable http client for retries and to honor throttle.
+// It also holds default HTTP headers to add to export requests.
+type otlpClient struct {
+	client  *retryablehttp.Client
+	headers map[string]string
+}
+
+// TelemetryClientCfg is used to configure the MetricsClient.
+type TelemetryClientCfg struct {
```

Comment:
Wondering if we need this struct as we only pass it to `NewMetricsClient` and we could just pass the fields directly? `func NewMetricsClient(cloudConfig, hclog.Logger) (MetricsClient, error)`

Reply:
will remove in upcoming commit - definitely overkill!